rust-src/curve_arithmetic/src/bls12_381_g1hash.rs

```diff
@@ -496,41 +485,139 @@ pub(crate) const K4: [[u64; 6]; 16] = [
+/// is a point on E' (written in Jacobian coordinates). Since inversions are
+/// expensive, we use the fact that (xn/xd : y : 1) on E' <=> (xn xd : xd^3y :
+/// xd) on E' when using Jacobian coordinates. It returns iso_11((xn xd : xd^3y
+/// : xd)) which is then guarentied to lie on E (but not necessarily in G1).
+fn map_to_curve(u: Fq) -> G1 {
+    let (xn, xd, mut y, _) = sswu_3mod4(u);
+    let mut x = xn;
+    x.mul_assign(&xd);
+    y.mul_assign(&xd);
+    y.mul_assign(&xd);
+    y.mul_assign(&xd);
+    let (xiso, yiso, z) = iso_11(x, y, xd);
+    from_coordinates_unchecked(xiso, yiso, z).expect("should not happen")
+}
+
+/// Implements https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#section-3
+/// with the choice of expand_message being expand_message_xmd, as specified in https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#section-8.8.1.
+fn hash_to_field(msg: &[u8], dst: &[u8]) -> (Fq, Fq) {
+    let (u_0, u_1, u_2, u_3) = expand_message_xmd(msg, dst);
+
+    (fq_from_bytes(&u_0, &u_1), fq_from_bytes(&u_2, &u_3))
+}
+
+// Interpret input as integers (big endian)
+// Return (left*2^256 + right) as Fq
+fn fq_from_bytes(left_bytes: &[u8; 32], right_bytes: &[u8; 32]) -> Fq {
+    fn le_u64s_from_be_bytes(bytes: &[u8]) -> Fq {
+        let mut digits = [0u64; 6];
+
+        for i in 0..4 {
+            digits[3 - i] = u64::from_be_bytes(
+                bytes[i * 8..(i + 1) * 8]
+                    .try_into()
+                    .expect("Chunk Size is always 8."),
+            );
+        }
+
+        Fq::from_repr(FqRepr(digits)).expect("Only the leading 4 u64s are initialized")
+    }
+
+    let two_to_256_fqrepr = [0u64, 0, 0, 0, 1, 0]; // 2^256
+    let two_to_256_fq = Fq::from_repr(FqRepr(two_to_256_fqrepr)).expect("2^256 fits in modulus");
+
+    let mut left_fq = le_u64s_from_be_bytes(left_bytes);
+    let right_fq = le_u64s_from_be_bytes(right_bytes);
+    left_fq.mul_assign(&two_to_256_fq); // u_0[..32] * 2^256
+    left_fq.add_assign(&right_fq); // u_0[32..] + u_0[32..] * 2^256 = u_0
+
+    left_fq
+}
+
+/// Implements https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#section-5.4.1
+/// len_in_bytes is fixed to 128
+/// Domain separation string (dst) should be at most 255 bytes
+fn expand_message_xmd(msg: &[u8], dst: &[u8]) -> ([u8; 32], [u8; 32], [u8; 32], [u8; 32]) {
+    // DST_prime = DST || I2OSP(len(DST), 1)
+    let mut dst_prime = dst.to_vec();
+    dst_prime.push(dst.len().try_into().unwrap()); // panics if dst is more than 255 bytes
+                                                   // msg_prime = Z_pad || msg || l_i_b_str || I2OSP(0, 1) || DST_prime
+
+    // b_0 = H(msg_prime)
+    let mut h = Sha256::new();
+    // todo a possible optimization here would be to save the state of H(Z_pad)
+    h.update(vec![0; 64]); // z_pad = I2OSP(0, 64), 64 is the input block size of Sha265
+    h.update(msg);
+    h.update(vec![0, 128]); // l_i_b_str = I2OSP(128, 2)
+    h.update(vec![0]);
```

Comment:
And here. I think it is also quite useful to document that `0` is a `u8` type, i.e., write `0u8` explictly.